app/models/deleted_object.rb

```diff
@@ -8,5 +8,10 @@ class DeletedObject < ApplicationRecord
     scope scoped_class.to_s.underscore.pluralize.to_sym, -> { where(object_type: scoped_class) }
   end
 
-  # TODO: cleanup regularly once the owner is destroyed too :) or this will grow uncontrollably
+  scope :service_owner, -> { where.has { owner_type == Service } }
+  scope :service_owner_not_exists, -> { service_owner.joining { owner.of(Service).outer }.where('services.id IS NULL') }
```

Comment:
It makes me think whether we need it always chained to the previous scope. I was thinking about something like `scope :missing_owner` that later could be chained `owned_by_service.missing_owner`. WDYT?

Reply:
@guicassolato That is what I tried first but I simply didn't find any way to make it work 🤷‍♀

Reply:
@guicassolato told me:

> i managed to solve for the first scope…

```ruby
scope :missing_owner, -> do
    associations = distinct(:owner_type).pluck(:owner_type).map(&:constantize).map do |association|
      -> { joining { owner.of(association).outer }.where("#{association.table_name}.id IS NULL") }
    end
    associations.inject(all) { |chain, association| chain.merge!(association) }
  end
```

> but still struggling with the events
> the problem is it could be any type of event of any type of object
> and `event_store_events` only holds a link to the associated object via an attribute inside the json data… when it does at all

Reply:
I said this:

> for the events i would also retrieve the kind of event and call to a class with that event type with the same name of scope for each and then:
> - if that scope exists (for example, let's call it `by_id`), then each of them adapts it to its own event type
> - if it doesn't exist is because it doesn't hold its id at all or because we didn't mind implementing it because we didn't need it, and in that case let's just make it as if it was a `none`